src/RouteLockedWorkspacesManager.php

```diff
@@ -0,0 +1,168 @@
+<?php
+
+namespace Drupal\workspaces_route_lock;
+
+use Drupal\Core\Cache\Cache;
+use Drupal\Core\Cache\CacheBackendInterface;
+use Drupal\Core\Entity\EntityTypeInterface;
+use Drupal\Core\Routing\RouteMatchInterface;
+use Drupal\workspaces\Entity\Workspace;
+use Drupal\workspaces\WorkspaceInterface;
+use Drupal\workspaces\WorkspaceManagerInterface;
+use Drupal\workspaces_route_lock\Entity\WorkspacesRouteLock;
+
+class RouteLockedWorkspacesManager implements WorkspaceManagerInterface {
+
+  /**
+   * The underlying manager service.
+   *
+   * @var \Drupal\workspaces\WorkspaceManagerInterface
+   */
+  protected $workspacesManager;
+
+  /**
+   * The current route match.
+   *
+   * @var RouteMatchInterface
+   */
+  protected $currentRouteMatch;
+
+  /**
+   * The cache backend.
+   *
+   * @var CacheBackendInterface
+   */
+  protected $cache;
+
+  /**
+   * WorkspacesModerationManager constructor.
+   *
+   * @param \Drupal\workspaces\WorkspaceManagerInterface $workspaceManager
+   *   The underlying manager service.
+   * @param \Drupal\Core\Routing\RouteMatchInterface $currentRouteMatch
+   *   The current route match.
+   * @param \Drupal\Core\Cache\CacheBackendInterface $cache
+   */
+  public function __construct(WorkspaceManagerInterface $workspaceManager, RouteMatchInterface $currentRouteMatch, CacheBackendInterface $cache) {
+    $this->workspacesManager = $workspaceManager;
+    $this->currentRouteMatch = $currentRouteMatch;
+    $this->cache = $cache;
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public function getActiveWorkspace() {
+    $activeWorkspace = $this->workspacesManager->getActiveWorkspace();
+    $routeName = $this->currentRouteMatch->getRouteName();
+
+    if ($routeName) {
+      $cid = "workspaces_route_lock:$routeName";
+      $cacheItem = $this->cache->get($cid);
+
+      if (!$cacheItem) {
+        /** @var \Drupal\workspaces_route_lock\Entity\WorkspacesRouteLockInterface $routeLock */
+        foreach (WorkspacesRouteLock::loadMultiple() as $routeLock) {
+          if ($routeLock->appliesToRoute($routeName)) {
+            if (!$routeLock->workspaceIsAllowed($this->getWorkspaceId($activeWorkspace))) {
+              $activeWorkspace = $routeLock->getWorkspaceToSwitchTo();
+              break;
+            }
+          }
+        }
+        $this->cache->set(
+          $cid,
+          $this->getWorkspaceId($activeWorkspace),
+          Cache::PERMANENT,
+          ['config:workspaces_route_lock_list']
+        );
+      } elseif ($cacheItem->data !== $activeWorkspace->id()) {
+        $activeWorkspace = $cacheItem->data;
+        if ($activeWorkspace) {
+          $activeWorkspace = Workspace::load($activeWorkspace);
+        }
+      }
+    }
+
+    return $activeWorkspace;
+  }
+
+  /**
+   * @inheritDoc
+   */
+  public function shouldAlterOperations(EntityTypeInterface $entity_type) {
+    return $this->isEntityTypeSupported($entity_type) && !$this->getActiveWorkspace()->isDefaultWorkspace();
```

Comment:
I've never used the decorator pattern with Drupal services. Yet I always wanted to try it because I thought that it should be better than the "classic" service class override.

But now I see the downsides.

Thanks :)

Reply:
You can swap the service only once, so if you do that, you basically become the dictator and no-one else is allowed to do changes to the service.

With decoration, however, many modules can alter the same service. It is harder to do, but also the way to go. Thanks to that workspaces route lock and https://www.drupal.org/project/workspace_moderation can work together for instance.